fix(semantic): allow reserved keywords in typescript ambient contexts

sapphi-red opened the pull request:
Reserved keywords are allowed as an identifier in ambient contexts.
```ts
declare const private: number
```
This is allowed.

sapphi-red commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15495?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15496** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15496?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#15495** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15495?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15495?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#15494** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15494?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#15493** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15493?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#15492** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15492?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=15495).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
> This probably should be solved in a different way...

The "look up ancestors" part only applies to `BindingIdentifier`s, right? If so, I think you can look up the `VariableDeclaration` with `Scoping::symbol_declaration` (or maybe it's not built yet?)

Dunqing commented:
I tweaked a little bit to improve the identifier in a BindingIdentifier situation, which already has a symbol ID, so that we can avoid getting a symbol ID again from `AstKind::BindingIdentifier`. I hope you don't mind that I modified the PR directly without conversation because I think the PR has been left here for a long time, so I don't want to delay it anyway!

sapphi-red committed:
fix(semantic): allow reserved keywords in typescript ambient contexts (#15495)

Reserved keywords are allowed as an identifier in ambient contexts.
```ts
declare const private: number
```
This is allowed.
